src/libolm_migration.rs

```diff
@@ -0,0 +1,446 @@
+    let device_id = account.device_id();
+    let identity_keys = &account.identity_keys;
+
+    let sessions = pickled_sessions
+        .into_iter()
+        .map(|pickled_session| {
+            Session::from_pickle(
+                user_id.to_owned(),
+                device_id.to_owned(),
+                identity_keys.clone(),
+                pickled_session,
+            )
+        })
+        .collect();
+
+    store.save_changes(Changes { sessions, ..Default::default() }).await?;
+    Ok(())
+}
+
+/// A pickled version of an `InboundGroupSession`.
+///
+/// Holds all the information that needs to be stored in a database to restore
+/// an InboundGroupSession.
+#[wasm_bindgen(getter_with_clone)]
+#[derive(Debug, Default)]
+pub struct PickledInboundGroupSession {
+    /// The pickle string holding the Megolm Session, as returned by
+    /// `olm_pickle_inbound_group_session` in libolm.
+    pub pickle: String,
+
+    /// The public curve25519 key of the account that sent us the session.
+    #[wasm_bindgen(js_name = "senderKey")]
+    pub sender_key: String,
+
+    /// The public ed25519 key of the account that sent us the session.
+    #[wasm_bindgen(js_name = "senderSigningKey")]
+    pub sender_signing_key: String,
+
+    /// The id of the room that the session is used in.
+    ///
+    /// Nullable so that a `PickledInboundGroupSession` can be constructed
+    /// incrementally. Must be populated!
+    #[wasm_bindgen(js_name = "roomId")]
+    pub room_id: Option<RoomId>,
+
+    /// Flag remembering if the session was directly sent to us by the sender
+    /// or if it was imported.
+    pub imported: bool,
+
+    /// Flag remembering if the session has been backed up.
+    #[wasm_bindgen(js_name = "backedUp")]
+    pub backed_up: bool,
+}
+
+#[wasm_bindgen]
+impl PickledInboundGroupSession {
+    /// Construct a new `PickledInboundGroupSession`, with default values.
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+#[wasm_bindgen]
+impl Migration {
+    /// Migrate Megolm sessions of a libolm-based setup to a vodozemac-based setup
+    /// stored in an indexedDB crypto store.
+    ///
+    /// Before this method can be used, {@link #migrateBaseData} must be used to
+    /// import the base data into the crypto store.
+    ///
+    /// # Arguments
+    ///
+    /// * `sessions` - An `Array` of {@link PickledInboundGroupSession}s to import.
+    /// * `pickle_key` - The libolm pickle key that was used to pickle the megolm
+    ///   session objects.
+    /// * `store_handle` - A connection to the CryptoStore which will be used to
+    ///   store the vodozemac data.
+    #[wasm_bindgen(js_name = "migrateMegolmSessions")]
+    pub async fn migrate_megolm_sessions(
+        sessions: Array,
+        pickle_key: Uint8Array,
+        store_handle: &StoreHandle,
+    ) -> Result<JsValue, JsError> {
+        let pickle_key = pickle_key.to_vec();
+
+        let rust_sessions = sessions
```

Comment:
If one of the megolm session to import is for some reason malformed/corrupted, will the whole chunk fail to import?
Maybe it would be better to ignore "malformed" and log something but not interrupt all?

Reply:
I'm not entirely convinced. This sounds like we could end up not migrating *any* sessions, and the user would just be confused.

Also: once the batch is successfully imported, we *delete* it from the legacy store. So, if we do fail to import any sessions, we'd end up completely throwing them away, which seems very dangerous.

I'd prefer to leave it and see if any errors actually arise in practice.

Reply:
Not exacty related but we have seen sort of similar things when [importing from backup](https://github.com/element-hq/element-web/issues/26774), also when migrating tracked users (legacy is more lax on userIds).

Could we return a result with the succesfully migrated things? so that the caller would delete only these and handle the error on the application side?
I'll let you decide what's best

Reply:
> Could we return a result with the succesfully migrated things? so that the caller would delete only these and handle the error on the application side?

Well, if we don't delete the broken sessions, then we'll just get stuck in a loop of repeatedly trying to migrate them. (There is no record of "which sessions have been migrated").)

I suppose one thing that we might think about is whether we can display an error to the user if migration goes wrong.